core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -370,24 +371,25 @@ class ContainerProxy(factory: (TransactionId, String, ImageName, Boolean, ByteSi
       .flatMap { activation =>
         val start = tid.started(this, LoggingMarkers.INVOKER_COLLECT_LOGS)
         collectLogs(tid, container, job.action)
-          .map { logs =>
-            tid.finished(this, start)
-            activation.withLogs(logs)
-          }
-          .recover {
-            case _ =>
+          .recoverWith {
+            case t =>
               tid.failed(this, start)
-              activation.withLogs(ActivationLogs(Vector(Messages.logFailure)))
+              val withLogs = t match {
+                case LogCollectingException(logs) => activation.withLogs(logs)
+                case _                            => activation.withLogs(ActivationLogs(Vector(Messages.logFailure)))
+              }
+              storeActivation(tid, withLogs)
+
+              Future.failed(ActivationLogReadingError(withLogs))
+          }
```

Comment:
It would be helpful to have a comment here that the `recoverWith` path with always end up in a failed Future so that the following `flatMap` will be skipped.